### PR TITLE
Use Azure Blob Storage for file uploads

### PR DESCRIPTION
**Pull request by epicfaace**
Diff: https://github.com/codalab/codalab-worksheets/compare/azurite..beam-uploads

For #2327.

Running `cl upload -a` will make that bundle be uploaded to Azure Blob Storage. The Blob Storage path will be stored in azfs://storageclwsdev0/bundles.

image

Done:
- uploading
- basic downloading
- support directories
- support using these bundles as dependencies

We don't support permissions yet; we can do that in a future PR.

### CODE REVIEW

File: codalab/lib/download_manager.py
Comment (percyliang) [inline codalab/lib/download_manager.py:311]: Maybe we should rename `_is_available_locally` since AzureFS is definitely not local (is available externally or directly)?

File: codalab/lib/download_manager.py
Comment (percyliang) [inline codalab/lib/download_manager.py:331]: Would be to refactor the check for `azfs://` in one place.  Also, let's think through what this will look like once we have `s3://`.  Probably makes sense to have a broader function that captures all these cases.

File: codalab/lib/path_util.py
Comment (percyliang) [inline codalab/lib/path_util.py:302]: Let's get in the habit of type annotating everything.  Should we make a dataclass instead of using a tuple?  Eventually, I think we should migrate all our tuples over to dataclasses, which will make the code easier to understand.

File: codalab/lib/path_util.py
Comment (percyliang) [inline codalab/lib/path_util.py:303]: Write a nicer docstring.

File: codalab/lib/upload_manager.py
Comment (percyliang) [inline codalab/lib/upload_manager.py:9]: What's this?  explain more?

File: codalab/lib/upload_manager.py
Comment (percyliang) [inline codalab/lib/upload_manager.py:66]: same comment as before about refactoring azfs

File: codalab/lib/zip_util.py
Comment (percyliang) [inline codalab/lib/zip_util.py]: Why not preserve the permissions now?

File: codalab/lib/zip_util.py
Comment (percyliang) [inline codalab/lib/zip_util.py:138]: Put a comment saying that we eventually want to unify this.

File: codalab/model/bundle_model.py
Comment (percyliang) [inline codalab/model/bundle_model.py:1055]: I don't think we will have a hash in the future since it would require reading everything, so just set it to None?

File: codalab/rest/bundles.py
Comment (percyliang) [inline codalab/rest/bundles.py:757]: What's our general solution here?  We should talk about how we want to name these for dev and prod?  I guess we should also document how these are setup in the deployment repo.

Comment (percyliang) [inline codalab/rest/bundles.py:757]: I thought we were just going to zip up everything for consistency and to simplify the logic?

File: codalab/server/bundle_manager.py
Comment (percyliang) [inline codalab/server/bundle_manager.py]: Why is this moved up?  In general, can you add a few more comments to your PRs?

File: codalab/worker/download_util.py
Comment (percyliang) [inline codalab/worker/download_util.py]: What's special about `/images/`?  Please put more comments, including adding a docstring.

File: codalab_service.py
Comment (percyliang) [inline codalab_service.py]: Is this only for testing though?  If so, then don't need this by default.

File: codalab/lib/bundle_cli.py
Comment (percyliang) [inline codalab/lib/bundle_cli.py]: I think our convention is to use `-` rather than `_` for all command-line args.

File: codalab/lib/beam/blobstoragefilesystem.py
Comment (percyliang) [inline codalab/lib/beam/blobstoragefilesystem.py:1]: Is this copied verbatim from somewhere?  If so, we should say where it's copied from, why we can't use it as a library, and any changes we're making.

File: tests/unit/lib/path_util_test.py
Comment (percyliang) [inline tests/unit/lib/path_util_test.py:141]: Refactor so that `storageclwsdev0` should only show up once, otherwise we'll have to change it in a lot of places.

File: tests/unit/server/bundle_manager/__init__.py
Comment (percyliang) [inline tests/unit/server/bundle_manager/__init__.py]: These bundle manager tests shouldn't be in this PR, right?

**Review by percyliang — CHANGES_REQUESTED** [review]
Exciting to see the Azure Blob storage come together!  I think my main comment is that we could refactor things a bit more out to set the stage for more general externalization.  Right now, Azure is hardcoded in a lot of places, and also, you kind of have to track the logic of how things are stored in a lot of places.  Could we consolidate this coupling behind a nice interface somehow?

File: codalab/lib/beam/filesystems.py
Comment (epicfaace) [inline codalab/lib/beam/filesystems.py:9]: @percyliang see this regarding where those two files (the copied verbatim files -- blobstoragefilesystem.py and blobstorageio.py) came from. Although that PR has been merged into beam, a new version hasn't yet been released with those changes, so we may just have to keep those two files in for now.

File: tests/unit/server/bundle_manager/__init__.py
Comment (epicfaace) [reply tests/unit/server/bundle_manager/__init__.py]: Forgot to mention this in the PR, but this PR is based on #2795, as I used some of the common utilities from that PR for testing here. It would be good to merge this once #2795 is merged.

File: codalab/server/bundle_manager.py
Comment (epicfaace) [reply codalab/server/bundle_manager.py]: See my comment about this PR being based on #2795 (this was a change done in #2795)

File: codalab/rest/bundles.py
Comment (epicfaace) [reply codalab/rest/bundles.py:757]: We need to pass the storage container name in as environment variables (will update to do that).

File: codalab/model/bundle_model.py
Comment (epicfaace) [reply codalab/model/bundle_model.py:1055]: Why not just keep it at `size` (which is what I think we discussed when originally designing --link)? I can remove the TODO comment if that's final, though.

File: codalab/lib/upload_manager.py
Comment (epicfaace) [reply codalab/lib/upload_manager.py:9]: Some of my comments / debugging statements are outdated 🙂 -- I'll update / remove them.

File: codalab/rest/bundles.py
Comment (epicfaace) [reply codalab/rest/bundles.py:757]: No, I think we decided against that -- we need to use `/contents` for single files, because otherwise we may not be able to distinguish between a bundle with a directory with a single file vs. a bundle with just a single file.

File: codalab/bundles/named_bundle.py
Comment (epicfaace) [inline codalab/bundles/named_bundle.py:27]: I needed to add these here so that we can have linked bundles that are non-UploadedBundles (we will need to do this anyway once run bundles upload their results to Azure Blob Storage).

File: codalab/lib/bundle_cli.py
Comment (epicfaace) [reply codalab/lib/bundle_cli.py]: Done

File: codalab/worker/download_util.py
Comment (epicfaace) [reply codalab/worker/download_util.py]: That was also for testing -- updated now.

File: tests/unit/lib/path_util_test.py
Comment (epicfaace) [reply tests/unit/lib/path_util_test.py:141]: Will do this for the general code, though for the tests themselves, it should be okay to use `storageclwsdev0` for now as we're just checking for parsing.

File: tests/unit/server/bundle_manager/__init__.py
Comment (epicfaace) [inline tests/unit/server/bundle_manager/__init__.py:66]: This class has effectively become a "test base class" generic test base that's used for non-BundleManager tests, so it should eventually be moved to a separate file. However, I'm holding off on doing that until #2795 is merged, in order to minimize merge issues / conflicts.

**Pull request closed without merging**
